I749 accessibility for graphs

pushyamig opened the pull request:
fixes #749

pushyamig commented:
This is still a Draft as what goes in the narrative still needs some feedback from the group. Feel free  to give any technical aspect of it

jonespm commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 5
           

Complexity increasing per file
==============================
- assets/src/components/d3/createHistogram.js  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ITS_Teaching_And_Learning/my-learning-analytics/pullRequest?prid=6313190&bid=20526236)

chrisrrowland commented:
I submitted a PR on your branch suggesting some changes that changed the cadence of how the text was read with my screen reader.  I'm curious how it affects yours, if at all.

chrisrrowland commented:
I've been testing keyboard accessibility by using tab to navigate.  I'm not sure if that it an appropriate way to do it or not, but it doesn't seem like the graph as as a whole is keyboard accessible?  I have to click on it to trigger the voice to text.

pushyamig commented:
when you switch between views, I can see that narrative being read. Sometime the trigger is tricky. When I talked to accessibility team about it the html semantics should be as described as [this](https://github.com/tl-its-umich-edu/my-learning-analytics/issues/749#issuecomment-702928710) for SVG.

chrisrrowland reviewed (approved):
Looks good to me

ssciolla commented:
Sorry for the delay, I'll try to review this in the next couple days, once I'm feeling better.

jonespm commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 12
           

Complexity increasing per file
==============================
- assets/src/util/chartsMisc.js  3
- assets/src/components/d3/createHistogram.js  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ITS_Teaching_And_Learning/my-learning-analytics/pullRequest?prid=6313190&bid=20526236)

jonespm commented on assets/src/components/d3/createResourceAccessChart.js:
```diff
@@ -146,12 +149,35 @@ function createResourceAccessChart ({ data, width, height, domElement }) {
+  // d3 graph narrative text
+  const narrativeTextResources = {}
+  narrativeTextResources.weekRange = `Resources accessed from week ${weekRange[0]} to ${weekRange[1]}.  `
+  narrativeTextResources.resourceType = `Selected resource type is ${resourceType}.  `
```
I'm looking for a course with multiple types of data what happens when you have more than one type selected?

jonespm commented on assets/src/components/d3/createResourceAccessChart.js:
```diff
@@ -146,12 +149,35 @@ function createResourceAccessChart ({ data, width, height, domElement }) {
+  narrativeTextResources.gradeFilter = gradeSelection.toUpperCase() !== 'ALL' ? `Filtering on grades ${gradeSelection}.  ` : ''
+  narrativeTextResources.resourcesUnaccessCount = `You have not yet accessed ${unReadResources.length} of ${resourceData.length} resources.  `
+  narrativeTextResources.resourcesList = resourceData.map(x =>
+    `${x.resource_name.split('|')[1]} of type ${x.resource_type} accessed ${x.total_percent}% and ${resourceToolTipText(x, false)} `
```
If we only have one type in the course or one type selected it's redundant to keep saying the resource_type every time. I feel like most classes will only have one time. This should be dropped out.

jonespm replied:
As a user I think it would be best if it had

The resourcesListUnaccessed
Then after the resourcesListAccessed

as two separate lists. It could read off all of the unaccssed entries then the accessed. I'm thinking of

This is good
`"Resources accessed from week 6 to 8.  Selected resource type is Files.  You have not yet accessed 4 of 36 resources."`

Then after
`<List of unacgessed resources without the resourceToolTipText, since it just says "You haven't accessed this resource">`

Then after that it could have
`"You have accessed 32 of 36 resources"`

And have the rest of the list of accessed resources. We might want to have a briefer resourceToolTipText for these cases?

pushyamig replied:
I have separated out the access vs unaccess narrative. Also I am not appending resource type if that is the selected type

jonespm reviewed (commented):
I have a few comments for improving the resources view but this is a good start.

I wonder if we could move this into a component, but it might just be the replacing or appending a div and concatenating the text together.

pushyamig commented:
> I wonder if we could move this into a component, but it might just be the replacing or appending a div and concatenating the text together.

I took the narrative text separate from the d3 view component making it more readable

jonespm commented:
The new changes look nice, I'll test this again locally but the code looks good.

pushyamig commented:
@jonespm is this ready for merge?

ssciolla commented on assets/src/components/d3/d3ViewsNarrative.js:
```diff
@@ -0,0 +1,57 @@
+import { isInRange, isOutOfRange } from '../../util/chartsMisc'
+
+const resources = (resourceData, resourceType, gradeSelection, weekRange) => {
```
Nit-pick: In my opinion, these functions (`resources`, `grades`) could be named better. I try to name functions with verbs (e.g. `createResourcesText`) because it helps me remember it's a function/method, though I understand in some cases they are used more like a property, which would make `resourcesText` okay. But when I see `resources` or `grades` alone (without the module name), it makes me think it is storing raw resource or grade data, which isn't true here. My two cents.

ssciolla reviewed (commented):
A couple things I saw when I glanced through this. I'll ask you directly if you want me to do a more complete review and test this.

ssciolla commented:
I'm also seeing there are some Codacy issues that we could review. Some of them can probably be ignored, but in my cursory glance, there are a couple others probably worth addressing.

pushyamig commented:
@ssciolla I did put you as a reviewer as well. If you have time you can review it

ssciolla commented on assets/src/components/d3/d3ViewsNarrative.js:
```diff
@@ -0,0 +1,57 @@
+  narrativeTextResources.gradeFilter = gradeSelection.toUpperCase() !== 'ALL' ? `Filtering on grades ${gradeSelection}.  ` : ''
+  narrativeTextResources.resourcesUnaccessCount = `You have not yet accessed ${unAccessedResources.length} of ${resourceData.length} resources.  `
+  narrativeTextResources.resourcesAccessCount = `${accesedResources.length !== 0 ? `. You have accessed ${accesedResources.length} of ${resourceData.length} resources.  ` : ''}`
+  narrativeTextResources.resourcesUnAccessList = unAccessedResources.map(x =>
+    `${resourceType.length === 1
+      ? `${x.resource_name.split('|')[1]} accessed ${x.total_percent}%`
+      : `${x.resource_name.split('|')[1]} of type ${x.resource_type} accessed ${x.total_percent}%`}`
+  )
+  narrativeTextResources.resourceAccessList = accesedResources.map(x =>
+    `${resourceType.length === 1
+      ? `${x.resource_name.split('|')[1]} accessed ${x.total_percent}% last time you accessed on ${new Date(x.self_access_last_time).toDateString()}`
+      : `${x.resource_name.split('|')[1]} of type ${x.resource_type} accessed ${x.total_percent}% last time you accessed on ${new Date(x.self_access_last_time).toDateString()}`}`)
```
I'm seeing the same thing here, I don't think that outer string formatting is necessary. It makes these blocks harder to read in my opinion.

ssciolla commented on assets/src/components/d3/d3ViewsNarrative.js:
```diff
@@ -0,0 +1,57 @@
+      : `${x.resource_name.split('|')[1]} of type ${x.resource_type} accessed ${x.total_percent}% last time you accessed on ${new Date(x.self_access_last_time).toDateString()}`}`)
+  return narrativeTextResources
+}
+const grades = (data, bins, gradesSummary, myGrade, firstGradeAfterBinnedGrade) => {
```
```suggestion

const grades = (data, bins, gradesSummary, myGrade, firstGradeAfterBinnedGrade) => {
```

ssciolla replied:
Nit-pick: This isn't a big deal, but I still think it would help readability to have a newline between `createResourcesText` and `createGradesText`.

ssciolla commented on assets/src/components/d3/d3ViewsNarrative.js:
```diff
@@ -0,0 +1,57 @@
+  const isBinningUsed = new Set(data.slice(0, 5)).size === 1
+  const binningNarrativeText = () => {
+    if (isOutOfRange(myGrade, firstGradeAfterBinnedGrade)) {
+      return `First grades are binned at ${Math.trunc(firstGradeAfterBinnedGrade)}% or lower and your grade ${myGrade} fall in this bin. `
+    } else {
+      return `First grades are binned at ${Math.trunc(firstGradeAfterBinnedGrade)}% or lower. `
```
I think this text could be clearer. Something like this?

"Grades at around {some score} or lower are placed into one bin, and your grade falls into this bin."

ssciolla commented on assets/src/components/d3/d3ViewsNarrative.js:
```diff
@@ -0,0 +1,57 @@
+    }
+  }
+  const userGradeInBinnedGroup = () => {
+    return isBinningUsed ? binningNarrativeText() : ''
```
I think I would just move `isBinningUsed ? binningNarrativeText() : ''` to line 37. This function doesn't seem necessary.

ssciolla commented on assets/src/components/d3/d3ViewsNarrative.js:
```diff
@@ -0,0 +1,57 @@
+  narrativeTextGrades.courseStats = `Course information: Class Size = ${gradesSummary.tot_students}, Average grade = ${gradesSummary.grade_avg}%, Median grade = ${gradesSummary.median_grade}%.`
+  narrativeTextGrades.binnedGradeText = userGradeInBinnedGroup()
+  narrativeTextGrades.courseGrades = []
+  for (const gradeBin in bins) {
```
Hm, I think `let` is more correct for a loop variable? Wonder why it lets you do that...

ssciolla replied:
`const` looks okay here, actually. Just did some googling, found this on SO: https://stackoverflow.com/questions/41067790/why-does-const-work-in-some-for-loops-in-javascript

ssciolla commented on assets/src/components/d3/d3ViewsNarrative.js:
```diff
@@ -0,0 +1,57 @@
+      const binLowerLimit = bins[gradeBin].x0
+      const binUpperLimit = bins[gradeBin].x1
+      if (isInRange(myGrade, binLowerLimit, binUpperLimit) && myGrade) {
+        narrativeTextGrades.courseGrades.push(`${bins[gradeBin].length} grades in ${binLowerLimit} - ${binUpperLimit}% range and your grade ${myGrade}% is in this range `)
```
I'd probably reword to "X grades are in the Y to Z range, and your grade is in this range."

ssciolla replied:
Also, see my other comments about the outer `${}`.

ssciolla commented on assets/src/util/chartsMisc.js:
```diff
@@ -0,0 +1,16 @@
+const isInRange = (data, lowerLimit, upperLimit) => {
+  if ((data > lowerLimit) && (data < upperLimit)) {
```
What about if the `data` === `lowerLimit` or `upperLimit`?

ssciolla commented on assets/src/util/chartsMisc.js:
```diff
@@ -0,0 +1,16 @@
+}
+
+const isOutOfRange = (data, checkPointData) => {
+  if (data !== null && data < checkPointData) {
```
Similar comment about `data === checkPointData`.

ssciolla commented on dashboard/views.py:
```diff
@@ -404,6 +403,8 @@ def resource_access_within_week(request, course_id=0):
             RESOURCE_VALUES.get(RESOURCE_VALUES_MAP.get(row.resource_type)).get('icon')
             ),
         axis=1)
+    # RESOURCE_VALUES_MAP {'canvas': 'files', 'leccap': 'videos', 'mivideo': 'videos'}
```
Should we keep this comment?

ssciolla commented on assets/src/components/d3/createHistogram.js:
```diff
@@ -44,9 +49,21 @@ function createHistogram ({
+    .attr('id', 'grade-view-narrative')
+    .attr('class', 'screenreader-only sr-only')
+    .text(d => {
+      return narrativeTextGrades.courseStats.concat(narrativeTextGrades.binnedGradeText, narrativeTextGrades.courseGrades.toString())
```
Yeah, I think it probably makes sense to move this concatenation into the `grades` function.

ssciolla reviewed (commented):
Nice work, @pushyamig, this is pretty cool. I'm sorry for all the comments. Most of them are not serious issues: they're about tidying up syntax, re-organizing some code so like logic is with like logic, tweaking language, etc. As usual, feel free to disregard issues you don't think need to be fixed. We should think more about those utility functions, to ensure they're covering the case of when grades are on range or bin boundary (i.e. should we be using `<=` or `>=`?

I still plan to test this, and I'll give my approval then if I don't see any other issues.

ssciolla commented on assets/src/components/d3/d3ViewsNarrative.js:
```diff
@@ -0,0 +1,57 @@
+  const accesedResources = resourceData.filter(resource => resource.self_access_count === 1)
+  const narrativeTextResources = {}
+  narrativeTextResources.weekRange = `Resources accessed from week ${weekRange[0]} to ${weekRange[1]}.  `
+  narrativeTextResources.resourceType = `Selected resource type is ${resourceType}.  `
```
What happens if there are multiple resource types ("Files" and "Videos" selected)? We may need more logic here.

ssciolla reviewed (commented):
Hi @pushyamig, I tested these, and it's working. I was able to find the narrative text with a screenreader (VoiceOver). In the process, I found a few other areas where the text might be improved, or where we need to consider some other cases (like the multiple resource types).

I'm not sure the `aria-live` attributes are working properly. I think you would want the narrative text to be automatically read if the visualization changes (due to a slider change or if "Show My Grade" is checked/un-checked), but I don't think that's triggering. I'm not sure what can be done about this. Seems tricky.

Again, I'm sorry for all the comments. I'm not intentionally trying to slow things down, these were just the things that came to mind while I was reviewing. I'm not going to approve yet, because I think there are a couple issues I've flagged that should be addressed (I'll update the corresponding comments to include "IMPORTANT"). The textual fixes could wait, though I don't think they would take terribly long to implement. Thanks for your patience, I hope this feedback feels valuable to you.

pushyamig commented on assets/src/components/d3/createHistogram.js:
```diff
@@ -44,9 +49,21 @@ function createHistogram ({
+  const main = d3.select(domElement).append('div')
+  // eslint-disable-next-line no-unused-vars
+  const containerGradesDiv = main.append('div')
+    .attr('aria-live', 'polite')
```
Done

pushyamig commented:
@ssciolla  I have resolved all the important issue and most of the other issues as well

ssciolla commented on assets/src/components/d3/d3ViewsNarrative.js:
```diff
@@ -0,0 +1,57 @@
+  const accesedResources = resourceData.filter(resource => resource.self_access_count === 1)
+  const narrativeTextResources = {}
+  narrativeTextResources.weekRange = `Resources accessed from week ${weekRange[0]} to ${weekRange[1]}.  `
+  narrativeTextResources.resourceType = `Selected resource type is ${resourceType}.  `
```
I think I saw when you demoed this that `resourceType` is actually an array? Or at the very least comma separated values. It'd be nice to make this grammatically correct. If you don't want to conditionally use "is" or "are", we could do something like "Selected resource type(s): ${resourceType}."

ssciolla commented on assets/src/util/math.js:
```diff
@@ -16,12 +16,26 @@ const median = arr => {
+}
+
+const isOutOfRange = (data, checkPointData) => {
+  if (data !== null && data <= checkPointData) {
```
Okay, so just checking here, I know I commented on this: `checkPointData` is the first value considered to be out of range? Or is the last value considered to be in range?

ssciolla commented on assets/src/components/d3/d3ViewsNarrative.js:
```diff
@@ -0,0 +1,57 @@
+  narrativeTextResources.resourceAccessList = accesedResources.map(x =>
+    `${resourceType.length === 1
+      ? `${x.resource_name.split('|')[1]} accessed ${x.total_percent}% last time you accessed on ${new Date(x.self_access_last_time).toDateString()}`
+      : `${x.resource_name.split('|')[1]} of type ${x.resource_type} accessed ${x.total_percent}% last time you accessed on ${new Date(x.self_access_last_time).toDateString()}`}`)
```
I'd still rate these suggestions as important. Not seeing that you changed them yet.

ssciolla replied:
Wups, sorry @pushyamig, maybe I was looking at an older commit. I see that you made the changes.

ssciolla commented on assets/src/components/d3/createHistogram.js:
```diff
@@ -44,9 +49,21 @@ function createHistogram ({
+  // SVG Components
+  const main = d3.select(domElement).append('div')
+  // eslint-disable-next-line no-unused-vars
+  const containerGradesDiv = main.append('div')
```
I noted that it is actually in the flow on Brave/Chrome. I'll mark this as resolved for now, but we may want to investigate more in the future, or avoid using `aria-hidden`?

ssciolla commented on assets/src/components/d3/d3ViewsNarrative.js:
```diff
@@ -0,0 +1,57 @@
+    return isBinningUsed ? binningNarrativeText() : ''
+  }
+  const narrativeTextGrades = {}
+  narrativeTextGrades.courseStats = `Course information: Class Size = ${gradesSummary.tot_students}, Average grade = ${gradesSummary.grade_avg}%, Median grade = ${gradesSummary.median_grade}%.`
```
This still seems significant to me, but it could be considered a separate issue about making that table more accessible (or at least making it produce more comprehensible screenreader output).

pushyamig commented:
> I'd still rate these suggestions as important. Not seeing that you changed them yet.

I did change as suggested 
```
 ? `${x.resource_name.split('|')[1]} has been accessed by ${x.total_percent}% of students. The last time you accessed this resource was on ${new Date(x.self_access_last_time).toDateString()}`
      : `${x.resource_name.split('|')[1]} of type ${x.resource_type} has been accessed by ${x.total_percent}% of students. The last time you accessed this resource was on ${new Date(x.self_access_last_time).toDateString()}`)```

Try looking in Files tab

pushyamig commented:
> > This still seems significant to me, but it could be considered a separate issue about making that table more accessible (or at least making it produce more comprehensible screenreader output).

Ok

ssciolla commented on assets/src/components/d3/d3ViewsNarrative.js:
```diff
@@ -0,0 +1,57 @@
+  narrativeTextResources.gradeFilter = gradeSelection.toUpperCase() !== 'ALL' ? `Filtering on grades ${gradeSelection}.  ` : ''
+  narrativeTextResources.resourcesUnaccessCount = `You have not yet accessed ${unAccessedResources.length} of ${resourceData.length} resources.  `
+  narrativeTextResources.resourcesAccessCount = accessedResources.length !== 0 ? `. You have accessed ${accessedResources.length} of ${resourceData.length} resources.  ` : ''
+  narrativeTextResources.resourcesUnAccessList = unAccessedResources.map(x =>
+    resourceType.length === 1
+      ? `${x.resource_name.split('|')[1]} accessed ${x.total_percent}%`
+      : `${x.resource_name.split('|')[1]} of type ${x.resource_type} accessed ${x.total_percent}%`
+  )
```
Okay, sorry, I see that you did make these changes in an earlier commit, but for some reason they are not being reflected in the code I pull down, or here in the cumulative changes. Can you please make sure they are there?

ssciolla commented on assets/src/components/d3/d3ViewsNarrative.js:
```diff
@@ -0,0 +1,57 @@
+
+const createResourcesText = (resourceData, resourceType, gradeSelection, weekRange) => {
+  const unAccessedResources = resourceData.filter(resource => resource.self_access_count === 0)
+  const accessedResources = resourceData.filter(resource => resource.self_access_count === 1)
```
```suggestion
  const accessedResources = resourceData.filter(resource => resource.self_access_count >= 1)
```

This is a bug. In cases where the `self_access_count` was greater than 1, no text was included for these resources.

ssciolla reviewed (approved):
Okay, thank you for your patience, @pushyamig. I'm approving, since I don't think you need to have me re-review again, but there are some suggestions I'm including here. Most of them are to further refine language or make the spacing/punctuation more consistent. There was also a bug in the `accessedResources` logic that needs to be fixed, but it's minor.
